zerver/migrations/0435_fix_attachment_realm.py

```diff
@@ -0,0 +1,68 @@
+import re
+
+from django.conf import settings
+from django.db import migrations
+from django.db.backends.base.schema import BaseDatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+
+
+def fix_attachment_realm(apps: StateApps, schema_editor: BaseDatabaseSchemaEditor) -> None:
+    Realm = apps.get_model("zerver", "Realm")
+    Attachment = apps.get_model("zerver", "Attachment")
+    ArchivedAttachment = apps.get_model("zerver", "ArchivedAttachment")
+
+    if not Realm.objects.exists():
+        return
+
+    internal_realm = Realm.objects.get(string_id=settings.SYSTEM_BOT_REALM)
+
+    print("")
+    for model in [Attachment, ArchivedAttachment]:
+        for attachment in model.objects.filter(realm_id=internal_realm.id).order_by("id"):
+            print(f"Fixing {model.__name__} {attachment.id}")
+            message = attachment.messages.only("realm_id").first()
+            if message is not None:
+                realm_id = message.realm_id
+            else:
+                # If attachment.messages is empty, try to infer the realm from path_id.
+                # The s3 backend set the path_id based on the correct realm, while the local
+                # storage backend formed path_id based on sender's realm (which was wrong
+                # for these attachments we're fixing, since the sender was a cross realm bot).
+                # We don't need to complicate the logic here with conditioning on the backend
+                # because worst case scenario, this just infers the realm to be the system bot
+                # realm, thus not changing anything at all.
+                matches = re.findall(r"^(\d+)\/", attachment.path_id)
+                if not matches:
+                    continue
+
+                try:
+                    realm_id = int(matches[0])
+                except ValueError:
+                    # Don't do anything if path_id doesn't start with a sensible realm id.
```

Comment:
Hmm yeah, if `Attachment`s like that exist and didn't get deleted then this would fail the migration. Tweaked with this to handle the edge case:
```
                    if not Realm.objects.filter(id=realm_id).exists():
                        # If the realm doesn't exist (e.g. due to deletion), we can't do anything.
                        continue
```